java/src/main/java/ai/rapids/cudf/JCudfSerialization.java

```diff
@@ -0,0 +1,412 @@
+        out.write(arrayBuffer, 0, amountToCopy);
+        bytesLeft -= amountToCopy;
+        byteOffset += amountToCopy;
+      }
+    } else {
+      int rshift = 8 - lshift;
+      while (bytesLeft > 0) {
+        int amountToCopy = (int) Math.min(bytesLeft, arrayBuffer.length - 1);
+        // Need to read at least 1 more byte to be sure we get any spill over.
+        column.copyHostBufferBytes(arrayBuffer, 0, ColumnVector.BufferType.VALIDITY, byteOffset, amountToCopy + 1);
+        byte currentByte = arrayBuffer[0];
+        for (int byteIndex = 0; byteIndex < amountToCopy; byteIndex++) {
+          byte nextByte = arrayBuffer[byteIndex + 1];
+          arrayBuffer[byteIndex] = (byte) ((nextByte << rshift) | ((0xFF & currentByte) >> lshift));
+          currentByte = nextByte;
+        }
+
+        out.write(arrayBuffer, 0, amountToCopy);
+        bytesLeft -= amountToCopy;
+        byteOffset += amountToCopy;
+      }
+    }
+    return padFor64bitAlignment(out, validityLen);
+  }
+
+  private static long copyAndPad(DataOutputStream out, ColumnVector column,
+                                       ColumnVector.BufferType buffer, long offset,
+                                       long length, byte[] tmpBuffer) throws IOException {
+    long left = length;
+    long at = offset;
+    while (left > 0) {
+      int amountToCopy = (int) Math.min(left, tmpBuffer.length);
+      column.copyHostBufferBytes(tmpBuffer, 0, buffer, at, amountToCopy);
+      out.write(tmpBuffer, 0, amountToCopy);
+      left -= amountToCopy;
+      at += amountToCopy;
+    }
+    return padFor64bitAlignment(out, length);
+  }
+
+  private static long copyBasicData(DataOutputStream out, ColumnVector column, long rowOffset,
+                                    long numRows, byte[] arrayBuffer) throws IOException {
+    DType type = column.getType();
+    long bytesToCopy = numRows * type.sizeInBytes;
+    long srcOffset = rowOffset * type.sizeInBytes;
+    return copyAndPad(out, column, ColumnVector.BufferType.DATA, srcOffset, bytesToCopy, arrayBuffer);
+  }
+
+  private static long copyStringData(DataOutputStream out, ColumnVector column, long rowOffset,
+                                     long numRows, byte[] arrayBuffer) throws IOException {
+    long startByteOffset = column.getStartStringOffset(rowOffset);
+    long endByteOffset = column.getEndStringOffset(rowOffset + numRows - 1);
+    long bytesToCopy = endByteOffset - startByteOffset;
+    long srcOffset = startByteOffset;
+    return copyAndPad(out, column, ColumnVector.BufferType.DATA, srcOffset, bytesToCopy, arrayBuffer);
+  }
+
+  private static long copyStringOffsets(DataOutputStream out, ColumnVector column, long rowOffset,
+                                        long numRows, byte[] arrayBuffer) throws IOException {
+    // If an offset is copied over as a part of a slice the first entry may be non-zero.  This is
+    // okay because we fix them up when they are deserialized
```

Comment:
Curious why the code decides to fix this up on the receiving end rather than on the sending end?  Seems like it would be more intuitive/readable to fix it up as the data is sent out rather than expecting bad offsets and then correcting them.

Reply:
We already are walking through the offsets on the CPU once when we receive it, why do it twice?  It is probably a needless micro-optimization and if you want me to switch it around I can.

Reply:
Oh and because DataOutputStream only writes values out as big endian, and depending on the system we are on we want the offsets to be in the native format so we can just do a memcopy I didn't want to have to try and figure out the current system endianness, I know we only support x86 right now but I believe power is also kind of supported, and potentially how to swap it, or write native code to do it I found it much simpler to do it the way it is now.